src/configurator/lib/playground-configurator.ts

```diff
@@ -0,0 +1,478 @@
+
+      <div id="rhs">
+        <div
+          id="container"
+          style="background-color:${this.values.getValue('pageBackground')}"
+        >
+          <code-sample
+            theme=${this.values.getValue('theme')}
+            .lineNumbers=${this.values.getValue('lineNumbers')}
+            project-src="./project/project.json"
+          >
+          </code-sample>
+
+          ${this.githubCorner}
+        </div>
+
+        <div id="code">
+          <div>
+            <h3>CSS</h3>
+            <codemirror-editor .value=${this.cssText} type="css" readonly>
+            </codemirror-editor>
+          </div>
+
+          <div>
+            <h3>HTML</h3>
+            <codemirror-editor .value=${this.htmlText} type="html" readonly>
+            </codemirror-editor>
+          </div>
+        </div>
+      </div>
+    `;
+  }
+
+  private get knobs() {
+    return knobSectionNames.map(
+      (section) =>
+        html`<section>
+          <h3 class="sectionLabel">${section}</h3>
+          <div class="knobs">
+            ${knobsBySection[section].map((knob) => this.knob(knob))}
+          </div>
+        </section>`
+    );
+  }
+
+  private get htmlText() {
+    return `
+<code-sample${this.htmlTextAttributes}>
+</code-sample>
+`;
+  }
+
+  private get htmlTextAttributes() {
+    const attributes = [];
+    for (const id of knobIds) {
+      const knob: Knob = knobsById[id];
+      if (!knob.htmlAttribute) {
+        continue;
+      }
+      const value = this.values.getValue(id);
+      if (value === knob.default) {
+        continue;
+      }
+      switch (knob.type) {
+        case 'checkbox':
+          if (value) {
+            attributes.push(` ${knob.htmlAttribute}`);
+          }
+          break;
+        default:
+          attributes.push(` ${knob.htmlAttribute}="${value}"`);
+          break;
+      }
+    }
+    return attributes.join('');
+  }
+
+  private get cssText() {
+    const props = [];
+    for (const id of knobIds) {
+      const knob: Knob = knobsById[id];
+      if (!(knob as Knob).cssProperty) {
+        continue;
+      }
+      const value = this.values.getValue(id);
+      let line = `${knob.cssProperty}: ${
+        knob.formatCss ? (knob as any).formatCss(value) : value
+      };`;
+      if (value === knob.default) {
+        line = `/*${line}*/`;
+      } else {
+        line = `  ${line}`;
+      }
+      props.push(line);
+    }
+    return `
+code-sample {
+${props.join('\n')}
+}
+    `;
+  }
+
+  private knob(knob: Knob) {
+    const label = html`<label for=${knob.id}>${knob.label}</label>`;
+    switch (knob.type) {
+      case 'select':
+        return [label, this.selectKnob(knob)];
+      case 'slider':
+        return [label, this.sliderKnob(knob)];
+      case 'color':
+        return [label, this.colorKnob(knob)];
+      case 'checkbox':
+        return [label, this.checkboxKnob(knob)];
+    }
+    return '';
+  }
+
+  private selectKnob(knob: KnobsOfType<'select'>) {
```

Comment:
Seems like these should be components... but I'm hesitant to suggest any investment before considering whether to use another library.

One reason _to_ develop a knob system here is to make hybrid source/inspector playgrounds where a sample can define its own knobs and relate them to source, so source and preview can change with the knob values. This could be used to build systems like the previews on material.io: https://material.io/components/buttons

Reply:
> Seems like these should be components... but I'm hesitant to suggest any investment before considering whether to use another library.

Yeah, I'd like to use Peter's knob system if it was open source. Agree I don't want to invest more in this, it works fine for what we need right now.

> One reason _to_ develop a knob system here is to make hybrid source/inspector playgrounds where a sample can define its own knobs and relate them to source, so source and preview can change with the knob values. This could be used to build systems like the previews on material.io: https://material.io/components/buttons

Interesting idea, will think about it.